### DISCUSSION TITLE
Ui thread and credential manager

### DISCUSSION THREAD

**Discussion by Arvant** [axmolengine/axmol#2231] [General]
Hi. Since the previous version of Google login was deprecated, I need to implement the new version. However, I'm facing an issue with the following function. When the user successfully signed in  or even makes a mistake, the game's UI no longer updates, and everything freezes. Does anyone have a solution for this?"

credentialManager.getCredentialAsync

[doc](https://developer.android.com/identity/sign-in/credential-manager#groovy)

public static void signInWithGoogle(final String txt) {
        GetCredentialRequest getCredRequest = new GetCredentialRequest.Builder()
            .addCredentialOption(googleIdOption)
            .build();

                credentialManager.getCredentialAsync(
                    // Use activity based context to avoid undefined
                    // system UI launching behavior
                    app,
                    getCredRequest,
                    null,
                    ContextCompat.getMainExecutor(app),
                    new CredentialManagerCallback<GetCredentialResponse, GetCredentialException>() {
                        @Override
                        public void onResult(GetCredentialResponse result) {
                            //handleSignInResult(result);
                        }

                        @Override
                        public void onError(GetCredentialException e) {
                            Log.i("Domino", "on errrrorrrrrrrrrr");
                        }
                    }
                );

    }

**Comment by johnspeny** [comment]
I have a working version, I noticed with this google login u need to log alot inorder to find issues.
This my version some parts are missing here. But see if it can help

```java

Context context = MainApplication.getInstance().getCurrentActivity();
        credentialManager = CredentialManager.create(context);

        CancellationSignal cancellationSignal = new CancellationSignal();
        cancellationSignal.setOnCancelListener(new CancellationSignal.OnCancelListener() {
            @Override
            public void onCancel() {
                Log.d("Java-App", "Getting credentials with Google was cancelled.");
                Toast.makeText(context, "Cancelled.", Toast.LENGTH_SHORT).show();
            }
        });
        Executor executor = Executors.newSingleThreadExecutor();


credentialManager.getCredentialAsync(context,
                    pendingGetCredentialHandle,
                    cancellationSignal,
                    executor,
                    new CredentialManagerCallback<GetCredentialResponse, GetCredentialException>() {
                        @Override
                        public void onResult(GetCredentialResponse result) {
                            handleSignIn(result );
                        }

                        @Override
                        public void onError(GetCredentialException e) {
                          
                            handleFailure(e);
                        }
                    });


 ```

**Reply by johnspeny** [reply]
[Don't forget to register your app in credentials page](https://developer.android.com/identity/sign-in/credential-manager-siwg#enable-sign-up)

**Reply by Arvant** [reply]
> I have a working version, I noticed with this google login u need to log alot inorder to find issues. This my version some parts are missing here. But see if it can help
> 
> ```java
> Context context = MainApplication.getInstance().getCurrentActivity();
>         credentialManager = CredentialManager.create(context);
> 
>         CancellationSignal cancellationSignal = new CancellationSignal();
>         cancellationSignal.setOnCancelListener(new CancellationSignal.OnCancelListener() {
>             @Override
>             public void onCancel() {
>                 Log.d("Java-App", "Getting credentials with Google was cancelled.");
>                 Toast.makeText(context, "Cancelled.", Toast.LENGTH_SHORT).show();
>             }
>         });
>         Executor executor = Executors.newSingleThreadExecutor();
> 
> 
> credentialManager.getCredentialAsync(context,
>                     pendingGetCredentialHandle,
>                     cancellationSignal,
>                     executor,
>                     new CredentialManagerCallback<GetCredentialResponse, GetCredentialException>() {
>                         @Override
>                         public void onResult(GetCredentialResponse result) {
>                             handleSignIn(result );
>                         }
> 
>                         @Override
>                         public void onError(GetCredentialException e) {
>                           
>                             handleFailure(e);
>                         }
>                     });
> ```

Thank you. I tested it, but it didn’t work for me. Did you run this code on the main AxmolActivity thread? I checked the sign-in process with logs. The user signed in successfully on Android, and I can get the user's Gmail address, but the UI does not update after that and freezes. I think the problem is with the threads. My onResume is called after sign-in, and touch works as well.

**Comment by Arvant** [comment]
i think problem is on axmolActivity @halx99:

```java
   @Override
    protected void onPause() {
    	Log.d(TAG, "onPause()");
        paused = true;
        super.onPause();
        AxmolEngine.onPause();
        mGLSurfaceView.onPause();
        mGLSurfaceView.setRenderMode(GLSurfaceView.RENDERMODE_WHEN_DIRTY);
    }

    @Override
    protected void onStop() {
        super.onStop();
        rendererPaused = true;
        mGLSurfaceView.onStop();
    } 
```

When I add rendererPaused = true; to the onPause() function, the problem is solved. @halx99, do you think I could create a pool for it, or would that cause any issues?

**Reply by rh101** [reply]
> When I add rendererPaused = true; to the onPause() function, the problem is solved. @halx99, do you think I could create a pool for it, or would that cause any issues?

By "pool", do you mean pull request?

More importantly, which version of Axmol are you using? That section of code has changed, so if you're not using v2.2.1 or higher, then please test that version to see if it solves your issue.  

Also, the suggested change would not work with the latest dev branch, since `rendererPaused` no longer exists.

**Reply by Arvant** [reply]
thank you my axmol version be old. after pull new version it work.

**Comment by Arvant** [comment]
Solved
